stackinator/builder.py

```diff
@@ -348,18 +348,95 @@ def generate(self, recipe):
+        repos = []
+
+        if recipe.spack_repo is not None:
+            self._logger.debug(f"adding recipe spack repo: {recipe.spack_repo}")
+            repos.append(recipe.spack_repo)
+
+        system_repo_path = system_config_path / "repo"
+        self._logger.debug(f"checking system spack repo: {system_repo_path}")
+        if spack_util.is_repo(system_repo_path):
+            self._logger.debug(f"adding system spack repo: {system_repo_path}")
+            repos.append(system_repo_path)
+
+        # look for repos.yaml file of relative paths....
+        repo_yaml = system_repo_path / "repos.yaml"
+        if repo_yaml.exists() and repo_yaml.is_file():
+            # open file
+            with repo_yaml.open() as fid:
+                raw = yaml.load(fid, Loader=yaml.Loader)
+                P = raw["repos"]
+
+            self._logger.debug(
+                f"the system spack repo contains {repo_yaml} which"
+                "refers to the following repos {P}"
+            )
+
+            # iterate
+            for rel_path in P:
+                repo_path = (system_config_path / rel_path).resolve()
+                if repo_path.is_dir():
+                    repos.append(repo_path)
+                    self._logger.debug(f"adding site spack repo: {repo_path}")
+
+        self._logger.debug(f"full list of spack repo: {repos}")
+
+        # Delete the target repo path, if it already exists.
+        # Do this so that incremental builds (though not officially supported)
+        # won't break if a repo is updated.
         repo_dst = store_path / "repo"
+        self._logger.debug(f"creating the stack spack repo in {repo_dst}")
         if repo_dst.exists():
+            self._logger.debug(f"{repo_dst} exists ... deleting")
             shutil.rmtree(repo_dst)
 
-        install(repo_src, repo_dst)
+        # If there are repos we iterate over them, copying their contents into the
+        # consolidated repo defined inside the uenv.
+        # We never overwrite a package that has already been installed, hence
+        # repos are in descending order of precidence.
+        if len(repos) > 0:
+            pkg_dst = repo_dst / "packages"
+            pkg_dst.mkdir(mode=0o755, parents=True)
+            self._logger.debug(f"created the repo packages path {pkg_dst}")
+            for repo_src in repos:
+                self._logger.debug(f"installing repo {repo_src}")
+                packages_path = repo_src / "packages"
+                for pkg_path in packages_path.iterdir():
+                    dst = pkg_dst / pkg_path.name
+                    if pkg_path.is_dir() and not dst.exists():
+                        self._logger.debug(
+                            f"  installing package {pkg_path} to {pkg_dst}"
+                        )
+                        install(pkg_path, dst)
+                    elif dst.exists():
+                        self._logger.debug(f"  not installing package {pkg_path}")
```

Comment:
```suggestion
                        self._logger.debug(f"  NOT installing package {pkg_path}")
```
Just to emphasize a bit... Maybe it would be worth adding the reason (i.e. "already included from a repo with higher precedence")?